src/osp/Active/drawing.h

```diff
@@ -24,67 +24,202 @@
+struct MaterialSkybox { };
+
+struct ACompMaterial
+{
+    material_id_t m_type;
+};
+
+//-----------------------------------------------------------------------------
+
 /**
- * A function pointer to a Shader's draw() function
- * @param ActiveEnt - The entity being drawn; used to fetch component data
- * @param ActiveScene - The scene containing the entity's component data
- * @param ACompCamera - Camera used to draw the scene
+ * @brief Stores a draw function and user data needed to draw a single entity
+ *
+ * This is usually generated by a DrawAssigner_t
  */
-using ShaderDrawFnc_t = void (*)(
-        ActiveEnt,
-        ActiveScene&,
-        ACompCamera const&) noexcept;
+struct EntityToDraw
+{
+    /**
+     * @brief A function pointer to a Shader's draw() function
+     *
+     * @param ActiveEnt    [in] The entity being drawn; used to fetch component data
+     * @param ActiveScene& [ref] The scene containing the entity's component data
+     * @param ACompCamera  [in] Camera used to draw the scene
+     * @param void*        [in] User data
+     */
+    using ShaderDrawFnc_t = void (*)(
+            ActiveEnt, ActiveScene&, ACompCamera const&, void*) noexcept;
+
+    ShaderDrawFnc_t m_draw;
+    // Non-owning user data passed to draw function, such as the shader
+    void *m_data;
+
+}; // struct EntityToDraw
 
 /**
- * Represents an entity which requests the scene to be drawn to m_target
+ * @brief Tracks a set of entities and their assigned drawing functions
+ *
+ * Each new entity added (in ACtxRenderGroups::m_added) is passed to one of
+ * every RenderGroup's DrawAssigner_t depending on its material type, which is
+ * responsible for setting its draw function and shader.
+ *
+ * DrawAssigners allow each RenderGroup to assign their own shaders for each
+ * kind of material. RenderGroups are intended to be associated with certain
+ * rendering techniques like Forward, Deferred, and Shadow mapping.
+ *
+ * This also works with game-specific modes like Thermal Imaging.
+ */
+struct RenderGroup
+{
+    using Storage_t = entt::storage_traits<ActiveEnt, EntityToDraw>::storage_type;
+    using ArrayView_t = Corrade::Containers::ArrayView<ActiveEnt>;
+
+    /**
+     * @brief Assigns drawing functions and user data (like shaders) to entities
+     *        of a certain material type for each RenderGroup
+     *
+     * @param ActiveScene& [ref] Scene the entities are part of
+     * @param Storage_t&   [ref] Storage to add accepted entities and data to
+     * @param ArrayView_t  [in] Entities to consider
+     */
+    using DrawAssigner_t = std::function<
+            void(ActiveScene&, Storage_t&, ArrayView_t) >;
+
+    /**
+     * @brief Set the DrawAssigner function for a certain material type
+     */
+    template<typename MATERIAL_T>
+    void set_assigner(DrawAssigner_t assigner)
+    {
+        material_id_t id = material_family_t::type<MATERIAL_T>;
+        m_assigners.resize(std::max(m_assigners.size(), size_t(id + 1)));
```

Comment:
skeptical of this.

If the type of MATERIAL_T is known at compile time (which it must be, given this is a template)

then unless you're planning to have dlls with previously unknown materials loaded at runtime, you should be able to determine the full list of possible materials at compile time, or at the very worst, at runtime, but during the static initalization phase prior to main() running, in the case of DLL injection, or at the very very very worst, after the DLL loader finishes, but before the game engine actually starts.

As a result, m_assigners should either be a std::array<> with that compile-time size, or it should be a std::unique_ptr<DrawAssigner_t[]> with the run-time determined size (either static initialization phase, or after all DLL loading completes)

Reply:
Also, are you checking to ensure that any particular item in the m_assigners vector is an initialized function prior to calling?

Reply:
It's possible to (easily?) modify entt::family to register new types at runtime from a config file or scripting engine. It would also be sane to just set to some large std::array (~256) since it will take a long time to reach that many material types. This function is just convenient, and it's performance isn't important since it's called only during setup.

They are checked in SysRender.cpp line 237:
```cpp
if ( ! bool(assigner) )
{
    continue; // Assigner is null, not set to a function
}
```

Reply:
Some details on how these materials will work: they are very broad 'suggestions' that are supported by components. They are suppose to be generic enough to never require specialization for individual objects.

ie. Instead of having ClearGlass, RedGlass, GreenGlass, and BlueGlass materials, there will only be one Glass material, and a color component. The assigner for glass will be responsible for reading the color, and setting the shader.